### ISSUE TITLE
Scan window is incredibly  "lagging"

### ISSUE THREAD

**Issue by ragulin72** [serratus/quaggaJS#450]
Hi,
I am using live stream for detection of barcodes via camera on the phone. 
High end mobile devices does not have a big problem.
Low / mid class mobile devices are struggling. The window in chrome App is lagging terribly. 
I have tried to set up the web workers, either to match the number of workers avalaible within the device, and setting it to 1 so it does not affect the performance. The results are though still same. 

Can you help me out? What can I do to get better performance out of it? 
Thank you very much for your help. 
My configuration looks like this:

```
    $(".controls").on("click", "button.start", function (e) {
        e.preventDefault();


        Quagga.init({
            inputStream: {
                name: "Live",
                type: "LiveStream",
                constraints: {
                    width: {ideal: 4096},
                    height: {ideal: 2160},
                    facingMode: "environment",
                    aspectRatio: {min: 1, max: 2}
                },
                area: {
                    top: "40%",
                    right: "20%",
                    left: "20%",
                    bottom: "40%"

                }

            },
            // frequency: 2,
            decoder: {
                readers: ["code_128_reader"],

            },
            locate: false,
            numOfWorkers:  self.navigator.hardwareConcurrency,
            locator: {
                patchSize: "x-small",
                debug: {
                    showCanvas: true,
                }
            }

        }, function (err) {
            if (err) {
                console.log(err);
                return
            }
            $(".controls").on("click", "button.stop", function (e) {
                e.preventDefault();
                Quagga.stop();
            });
            Quagga.start();

            var track = Quagga.CameraAccess.getActiveTrack();
            console.log (track);
            Quagga.onProcessed(function (result) {
                var drawingCtx = Quagga.canvas.ctx.overlay,
                    drawingCanvas = Quagga.canvas.dom.overlay;

                if (result) {
                    if (result.boxes) {
                        drawingCtx.clearRect(0, 0, parseInt(drawingCanvas.getAttribute("width")), parseInt(drawingCanvas.getAttribute("height")));
                        result.boxes.filter(function (box) {
                            return box !== result.box;
                        }).forEach(function (box) {
                            Quagga.ImageDebug.drawPath(box, {x: 0, y: 1}, drawingCtx, {color: "green", lineWidth: 2});
                        });
                    }
                    if (result.codeResult && result.codeResult.code) {
                        Quagga.ImageDebug.drawPath(result.line, {x: 'x', y: 'y'}, drawingCtx, {color: 'red', lineWidth: 3});
                    }
                }
            });
            Quagga.onDetected(function (result) {
                var code = result.codeResult.code;
                if (Quagga.lastResult !== code) {
                    console.log(result.codeResult);
                    Quagga.lastResult = code
                    console.log(code);
                    var audio = new Audio("../files/barcode-beep_1.mp3")
                    audio.play()
                    var $node = null, canvas = Quagga.canvas.dom.image;
                    // navigator.vibrate(300);
                    document.getElementById("frm-searchForm-op").value = code;
                    document.getElementById('frm-searchForm-op-label').classList.add('active')

                }
            });
        })
    })```

**Comment by ericblade** [comment]
You might want to try the "Performance" tab in the Chrome Inspector.  I don't see anything obvious and horrible that you're doing there, although i would suspect that if you're actually getting a 4Kx2K camera resolution, you're probably working with *much* larger images than were even possible on a mobile at the time that much of quagga was developed :-)   That could cause a slowdown.  Patches being in x-small with a huge image might also be an issue. 

I'd like to find out what Performance says, I definitely don't have access to any hardware capable of that kind of an image, so i'd be curious.

**Comment by ragulin72** [comment]
> You might want to try the "Performance" tab in the Chrome Inspector. I don't see anything obvious and horrible that you're doing there, although i would suspect that if you're actually getting a 4Kx2K camera resolution, you're probably working with _much_ larger images than were even possible on a mobile at the time that much of quagga was developed :-) That could cause a slowdown. Patches being in x-small with a huge image might also be an issue.
> 
> I'd like to find out what Performance says, I definitely don't have access to any hardware capable of that kind of an image, so i'd be curious.

I am sorry for late reply.
Well from what I know, the note "ideal" in the code defines it to use the best possible resolution. The device I test on has 2K res, but it is Samsung J3 with low CPU. 
On Iphone 8+ it is not lagging at all.
I am trying to make it work smoothly at low end devices like Samsung J3 as users will work mostly with devices like this. 

Im a bit lost in the performance tab, not sure what shall I look for. May I somehow export the report and post it over here?

**Comment by ericblade** [comment]
right, if you're telling it "ideal" is 4kX2k, and whatever you're running on actually has the ability to give you an image that big, you're going to spend a lot of time processing images.  

Honestly, I haven't really used the modern Performance tab in quite some time, I don't really know.  I'm just going to turn on some profiling on this github tab, and start typing in and see what happens

**Comment by ericblade** [comment]
hmm.  not seeing any obvious hotspots in github to really give me an idea of what 'bad' looks like here.  i think if you turn on performance capturing while you're experiencing what looks like a slowdown, you can probably locate some places in the graph that comes up, that are running slower than 'normal' and drill down into the Call Tree tab below the graph to see what code is running during the slowdowns.

I would also try turning down the resolution on devices that are giving problems, and see if the problem goes away.

**Comment by ragulin72** [comment]
> hmm. not seeing any obvious hotspots in github to really give me an idea of what 'bad' looks like here. i think if you turn on performance capturing while you're experiencing what looks like a slowdown, you can probably locate some places in the graph that comes up, that are running slower than 'normal' and drill down into the Call Tree tab below the graph to see what code is running during the slowdowns.
> 
> I would also try turning down the resolution on devices that are giving problems, and see if the problem goes away.


Well thank you for advice. From the Call Tree, there is nothing what would indicate problems with performance. The most time taken is by 89% "Animation frame fired". 

The application is laggy if I leave the maximal resolution possible for device.

If I set up strictly 1024x768, it is lagging less. 

Problem is with recognition of some barcodes (dependent more on the place where the barcode is, than on its shape or size).

So the result would in my eyes be:
Quagga does not work well on low end devices while high resolution is used for the live stream from camera.

Frustrating fact is, that paid solutions like barcode scanner from Cognex work almost "Immidieatly" and without any problems at any kind of barcode or whatever. I ask myself wether their code is so much better and complex than possibilities of Quagga, or if just Im bad programmator.

**Comment by ericblade** [comment]
well, there's a couple of things that i've seen that people have done to cause performance issues before -- one is just running a lot of code inside your onDetected or onProcessed callbacks.  Another is adding multiple instances of your onDetected or onProcessed callbacks.  There was also a problem in my fork of this for a little while, where I was failing to stop the animation timer at stop(), but starting a new one at start(), so every time the user clicked stop/start it would start running detection X times per frame, which was a huuuge cpu lag.  

You should be able to acceptably use a lower camera resolution usually -- i believe i default to 640x480 where possible, and i think i had to use minimum of 800x600 on iOS devices, and it usually works pretty well for me.

A couple of big differences between Quagga and other solutions, is that Quagga is running directly in your browser, in Javascript .. while other solutions are native code .. and also Quagga has some pretty advanced functionality for detecting where in the image and what orientation a barcode is in.  

You can reduce CPU time by turning off the barcode locator function, reducing the rate of sampling, and increasing the patch sizes.  There may be other options that affect it as well.  Make sure you're only using the number of readers that you need, and try with numOfWorkers 0.  Ultimately, though, scanning huge images takes time.

Here's what a bit of call stack looks like on my application while the camera is running

![image](https://user-images.githubusercontent.com/1451847/99700868-bca42380-2a61-11eb-9ac1-b65e460e7633.png)

1558ms is a *lot* of time to be spent, and this is on my laptop, which is pretty beefy compared to most cell phones.  

The biggest chunk of time there is spent in locate(), it might be possible that there's some optimizations that might be able to be done in there, but i don't really know how any of that code works and i'm sort of afraid to mess with it :)   grab() is another big spike there, but i don't think there's much that can be done there.  

One thing I do see is that there's a lot of unexpected garbage collection occurring, so that might be something to look into on my fork, and see if i can reduce how much that's happening.

**Comment by manesvenom** [comment]
Is it possible to limit the scan area within a preset boundary and disable the locate function to improve the performance for advanced setting or only apply the locate function  on the preset boundary.
![IMG_BF64516026D5-1](https://user-images.githubusercontent.com/7610019/106544516-a119c680-6542-11eb-82dc-968c2176a651.jpeg)

I have tried a demo by this company using the browser for scanning as well, their speed is fast.
https://www.dynamsoft.com/Products/barcode-recognition-javascript.aspx

**Comment by ericblade** [comment]
set locate: false and define an area, per the readme

it does make it faster/less intensive, but does reduce the usability, imo
